src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -254,6 +271,465 @@ create_distributed_table(PG_FUNCTION_ARGS)
+	 * we want to error with a user friendly message if foreign keys are not supported.
+	 * We can miss foreign key violations because we are not holding locks, so relation
+	 * can be modified until we acquire the lock for the relation, but we do as much as we can
+	 * to be user friendly on foreign key violation messages.
+	 */
+	EnsureForeignKeysForDistributedTableConcurrently(relationId);
+
+	/*
+	 * Get name of the table before possibly replacing it in
+	 * citus_add_local_table_to_metadata.
+	 */
+	char *tableName = get_rel_name(relationId);
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	RangeVar *rangeVar = makeRangeVar(schemaName, tableName, -1);
+
+	/*
+	 * Before taking locks, convert the table into a Citus local table and commit
+	 * to allow shard split to see the shard.
+	 */
+	SendAddLocalTableToMetadataCommandOutsideTransaction(relationId);
+
+	/*
+	 * Lock target relation with a shard update exclusive lock to
+	 * block DDL, but not writes.
+	 *
+	 * If there was a concurrent drop/rename, error out by setting missingOK = false.
+	 */
+	bool missingOK = false;
+	relationId = RangeVarGetRelid(rangeVar, ShareUpdateExclusiveLock, missingOK);
+
+	if (PartitionedTableNoLock(relationId))
+	{
+		/* also lock partitions */
+		LockPartitionRelations(relationId, ShareUpdateExclusiveLock);
+	}
+
+	List *shardList = LoadShardIntervalList(relationId);
+
+	/*
+	 * It's technically possible for the table to have been concurrently
+	 * distributed just after citus_add_local_table_to_metadata and just
+	 * before acquiring the lock, so double check.
+	 */
+	if (list_length(shardList) != 1 ||
+		!IsCitusTableType(relationId, CITUS_LOCAL_TABLE))
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("table was concurrently modified")));
+	}
+
+	/*
+	 * The table currently has one shard, we will split that shard to match the
+	 * target distribution.
+	 */
+	ShardInterval *shardToSplit = (ShardInterval *) linitial(shardList);
+
+	PropagatePrerequisiteObjectsForDistributedTable(relationId);
+
+	bool viaDeprecatedAPI = false;
+	char replicationModel = DecideReplicationModel(distributionMethod,
+												   colocateWithTableName,
+												   viaDeprecatedAPI);
+
+	Var *distributionColumn = BuildDistributionKeyFromColumnName(relationId,
+																 distributionColumnName,
+																 NoLock);
+
+	Oid distributionColumnType = distributionColumn->vartype;
+	Oid distributionColumnCollation = distributionColumn->varcollid;
+
+	/*
+	 * At this stage, we only want to check for an existing co-location group.
+	 * We cannot create a new co-location group until after replication slot
+	 * creation in NonBlockingShardSplit.
+	 */
+
+	/* get an advisory lock to serialize concurrent default group creations */
+	AcquireColocationDefaultLock();
+
+	bool colocatedWithDefault = false;
+	uint32 colocationId = FindColocateWithColocationId(relationId,
```

Comment:
FindColocateWithColocationId has some useful errors checks, e.g. it checks the distribution column type of the colocate_with table. It would be nice if we could also do this (on a best effort basis) before the Citus local table conversion.